### ISSUE TITLE
Interoperable Private Attribution (IPA)

### ISSUE THREAD

**Issue by eriktaubeneck** [patcg/private-measurement#9]
@benjaminsavage, @martinthomson, and I have been working on a proposal, "[Interoperable Private Attribution (IPA)](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit#)" that addresses the aggregate attribution measurement use case, similar to those listed in patcg/private-measurement#8. 

We'd love to have this considered and discussed at the January PATCG meeting, for consideration in maturing it further through collaboration among this community group.

**Comment by csharrison** [comment]
I reread the privacy budgeting section of IPA and I have a small concern. It seems like the proposed privacy "grain" / unit is advertiser x user but we implement that via a grain of advertiser x match key. However, match keys and users are not the same and this seems like it is possibly abusable.

For a simple example say you are targeting a single sensitive user across 3 devices. The identity provider could intentionally use a separate match key for each device, and send separate queries for each match key that consume budgets independently for each query, with the intention of leaking as much about this user as possible. Note that by doing this we sort of eliminate the possibility of true cross-device attribution though.

Making match keys easy to swap out on a single device (possibly across apps) makes this attack worse, but that seems fixable with OS-level support. Best case I can see is that this proposal would achieve (in the worst case) is site x user x device privacy.

LMK if I am missing something here though.

**Comment by csharrison** [comment]
> Making match keys easy to swap out on a single device (possibly across apps) makes this attack worse, but that seems fixable with OS-level support. Best case I can see is that this proposal would achieve (in the worst case) is site x user x device privacy.

Actually, I am not sure this is easy. I think we would need to make sure that two "different entities" (e.g. facebook.com and facebook2.com) don't try to measure the same events with two different match keys (user123-A and user123-B), to double up their counts. In other words I think match key consistency needs to be 1:1 with websites.

**Comment by martinthomson** [comment]
This is an attack we've discussed (though we might not have captured it in documentation).  You need two controls:

1. The user agent cannot act on a change to a match key until the start of a new epoch.  (Other rate limiting is possible, but this is easiest.)  This prevents the identifier from being swapped out on a per-request basis.
2. Sites need to commit to a set of match key providers so that sites can't switch out providers in the same way (to your second point).  It might be acceptable here for each user agent to independently enforce this, though this means that you really end up with (user+device/agent) x site as your grain.  Ideally, some sort of consistency system would be used to get back a true user+site grain; it's possible the helpers can play some role in guaranteeing that.

**Comment by csharrison** [comment]
Yes (1) makes sense and I think it's specified in the doc. (2) doesn't fully cover the second comment though. I could ask all my partner sites to pre-commit to use fb1.com / fb2.com as identity providers without doing any kind of rotation / switching. Then I just double up all my source / trigger events (one per each match key provider) and get effectively double counts for the same privacy budget.

I think what you might have to do (and maybe this is what you were implying with (2)) is that if you pre-commit to N match key providers, you would effectively have more noise added to your aggregations scaled with N to keep similar privacy guarantees and get closer to the user+site grain.

**Comment by eriktaubeneck** [comment]
In the documentation, we originally proposed providing `match_key_provider` as an argument in the `generateSourceEvent` and `generateTriggerEvent` functions. However, this is vulnerable to the attack to describe @csharrison.

@martinthomson's suggestion of committing to a set `{match_key_provider_1, ...}` would be for every site. Then, when calling `generateSourceEvent` or `generateTriggerEvent`, all of the match keys that site had committed to would be included. (We'd want to be able to cap this to some reasonable number `N`.)

Later, in the privacy budget management step, we'd assure that both:
1. Every individual match key can only contribute up to `L1` to the aggregation.
2. Every individual match key provided has the consumed amount deduced from their privacy budget.
  
I believe this should allow for properly preventing the attack you describe, without needing to scale the noise by a factor of `N`.

**Comment by benjaminsavage** [comment]
> I could ask all my partner sites to pre-commit to use fb1.com / fb2.com as identity providers without doing any kind of rotation / switching. Then I just double up all my source / trigger events (one per each match key provider) and get effectively double counts for the same privacy budget.

Like @eriktaubeneck says, I think the idea is that if you want to use multiple match-key providers, that's fine, but each event will contain ALL of them. You don't get to pick and choose. I think that eliminates this attack.

If you ask all your partner sides to pre-commit to use fb1.com / fb2.com as identity providers, then each time they generate an IPA event, it'll use both match keys (if present, defaulting to a randomly generated but stable-per-device value if not present).

**Comment by benjaminsavage** [comment]
You're totally right @csharrison. My thinking is that given each site owner can independently choose which match-key providers they work with, any provider who does something nasty like choosing a uniform value of the match key, will instantly lose trust and develop a bad reputation - leading to nobody using them again.

**Comment by eriktaubeneck** [comment]
@csharrison another approach would be to structure the inclusion of matchkeys in the report as a key:value of provider:matchkey, instead of just a set of matchkeys, i.e.

```
{
    "provider1.com": matchkey_1,
    "provider2.com": matchkey_2,
    ...
}
```

Then, at query time, you could tell the aggregators: "only join on provider1.com". The aggregators could respect that choice in the joining, but still account for budgeting against the full set of matchkeys.

This still wouldn't fully solve the abuse scenario, however, because if someone were to simply set a uniform matchkey, that would likely still disrupt the budget accounting and contribution capping (in which case you'd still need reputational effects which @benjaminsavage proposes.)

**Comment by csharrison** [comment]
Another question for the IPA proposal. The document mentions it should be possible for third parties to make requests on behalf of other 1Ps. I agree this is a good feature. One attack I didn't see mentioned is malicious parties crafting fake data in the hope of stealing budget from the 1P, by pretending to query on behalf of the 1P.

There are many mitigations for this, but it would be good to spell them out. The most obvious one is that if the match key space is high entropy enough, this is just straight up difficult. However, I don't know if we want to design something more robust such that e.g. 1Ps need to attest to working with certain 3Ps up front.

**Comment by eriktaubeneck** [comment]
@csharrison agreed that this is underspecified, and this would be a great area to get more clarity on. 

[Administrative side note, I opened a [request](https://github.com/patcg-individual-drafts/admin/issues/1) to get a repo specifically for IPA so we can have issues to dedicated topics, and even put together pull requests for docs outline more details in these areas as they emerge.]

A few thoughts specific to to this question:

> One attack I didn't see mentioned is malicious parties crafting fake data in the hope of stealing budget from the 1P, by pretending to query on behalf of the 1P.

In the case where the 3P has actual `source_events` and `trigger_events`, this could be possible without even generating fake data. We allow for individual events to be used in multiple queries, within the privacy budget, so this could be used to exhaust it. In this case, I don't think that making the match key space high entropy would actually work.

In the case where the 3P doesn't have actual events, but is just trying to disrupt some 1P's budget, the high entropy match key space would work. 

> I don't know if we want to design something more robust such that e.g. 1Ps need to attest to working with certain 3Ps up front.

In the first scenario, it should be possible for the 1P to prevent a 3P from getting actual events by not sending them or installing their "pixel" code.

In the second scenario, it seems like a high entropy match key is enough (say 64-bit) where it would be far too expensive to run a query that would actually have meaningful impact. Let's suppose (very conservatively) that it only takes 1ms to generate a fake event - to cover 0.4% (1/256) of the space it would take over 2M years of compute time to generate all those events. And that's not even starting to think about actually running that query...

That said, if a 1P wants to work with more than one 3P, then we do probably need a way for that 1P to assign specific portions of its budget across those different 3Ps, which may necessitate the attestation design you mention.

**Comment by csharrison** [comment]
> [Administrative side note, I opened a https://github.com/patcg-individual-drafts/admin/issues/1 to get a repo specifically for IPA so we can have issues to dedicated topics, and even put together pull requests for docs outline more details in these areas as they emerge.]

Thanks, yeah. This one issue is getting very cumbersome haha.

> In the first scenario, it should be possible for the 1P to prevent a 3P from getting actual events by not sending them or installing their "pixel" code.

It's worth thinking through this scenario to see if we could detect / tolerate this. As far as I understand things, it is notoriously difficult for 1Ps to make configuration changes on their sites, so if we are relying on that to deter cheaters it's not ideal.

> In the first scenario, it should be possible for the 1P to prevent a 3P from getting actual events by not sending them or installing their "pixel" code.

Hm, this made me look back at the IPA doc to see how privacy budgeting is done. For a given report in the MPC system, how do we know the site it is associated with for purposes of budget? The issue I am hoping we can avoid is something like a re-randomization attack where a 3P gets _real_ events for advertiser A but can somehow use the match key to steal budget from advertiser B while having the report look new. I think we need to make sure the budget keys are not tamperable basically.

I think I agree with you about the high entropy protecting us a great deal from the "guessing" attack. If we can show that's the worst an adversary can do I might be comfortable with it.

**Comment by eriktaubeneck** [comment]
> It's worth thinking through this scenario to see if we could detect / tolerate this. As far as I understand things, it is notoriously difficult for 1Ps to make configuration changes on their sites, so if we are relying on that to deter cheaters it's not ideal.

I agree if we're talking about a cheater using information from site A to impact something about site B. However, if site A is willing to give a "cheater" ability to execute JS on their site, how we can prevent anything beyond that.

> For a given report in the MPC system, how do we know the site it is associated with for purposes of budget?

This is a good question. I have a few ideas here, but there are some tradeoffs. I'll open an issue for this specifically once the other repo is created.

**Comment by csharrison** [comment]
> I agree if we're talking about a cheater using information from site A to impact something about site B. However, if site A is willing to give a "cheater" ability to execute JS on their site, how we can prevent anything beyond that.

Great point. There might be some nuance here with iframes, but even still there is a detection problem that we should think through. This goes back to the general problem of supporting multiple separate reporting origins though which we should flesh out. It ends up being a complicated coordination problem (and possible denial-of-service vector) if everyone has to share a single budget. Needing to involve the advertiser in it makes this even tougher.

One more question about IPA behavior. I want to confirm that attribution across multiple queries works correctly. Here's an example: imagine there's a single advertiser selling a single product, and is running 3 different campaigns for it. They want to use IPA to measure the relative performance of their 3 campaigns. This is my understanding of how this would work in IPA.

The advertiser will send 3 queries to the system:
1. {Campaign1's source events, all trigger events}
2. {Campaign2's source events, all trigger events}
3. {Campaign3's source events, all trigger events}

If IPA treats these queries completely independently, then attribution does not take into account source events from separate queries. That is, a hypothetical user journey like {Campaign1 source event, Campaign2 source event, Campaign 3 source event, trigger} will end up contributing a count to each of the three queries above, causing double counting.

One way to make this work would be to first run "global attribution" with the union of the events in all the queries, and then separately evaluate each query separately from the pool of globally attributed sources/triggers. I couldn't tell if this was how the protocol was intended to work though.

**Comment by benjaminsavage** [comment]
> Another question for the IPA proposal. The document mentions it should be possible for third parties to make requests on behalf of other 1Ps. I agree this is a good feature. One attack I didn't see mentioned is malicious parties crafting fake data in the hope of stealing budget from the 1P, by pretending to query on behalf of the 1P.

Here's how I've been thinking about this:

When a report collector makes an IPA query, it will cost them some amount of money. You have to pay the MPC helper nodes for the compute you use. This implies the existence of some kind of registration process whereby a site / app signs up to run IPA queries, proves ownership of the app / site, and inputs an associated payment instrument. 

So I am assuming all IPA queries will be authenticated server-to-server calls. Authentication parameters must be provided to run the query. As such, it should be impossible for anyone but the 1st party, or their legitimate delegate to run queries. If a delegate abuses their permissions, the 1st party should be able to revoke their permission to run IPA queries on their behalf.

**Comment by benjaminsavage** [comment]
> One more question about IPA behavior. I want to confirm that attribution across multiple queries works correctly. Here's an example: imagine there's a single advertiser selling a single product, and is running 3 different campaigns for it. They want to use IPA to measure the relative performance of their 3 campaigns. This is my understanding of how this would work in IPA.
> 
> The advertiser will send 3 queries to the system:
> 
> {Campaign1's source events, all trigger events}
> {Campaign2's source events, all trigger events}
> {Campaign3's source events, all trigger events}
> If IPA treats these queries completely independently, then attribution does not take into account source events from separate queries. That is, a hypothetical user journey like {Campaign1 source event, Campaign2 source event, Campaign 3 source event, trigger} will end up contributing a count to each of the three queries above, causing double counting.
> 
> One way to make this work would be to first run "global attribution" with the union of the events in all the queries, and then separately evaluate each query separately from the pool of globally attributed sources/triggers. I couldn't tell if this was how the protocol was intended to work though.

In the event an advertiser wants to evaluate the relative performance of 3 campaigns (which they might have purchased from different ad-sellers) I assume that they would NOT issue three separate queries as you’ve shown. This would wind up hitting their differential privacy budget three times for the same set of trigger events. They’d be far better off running a single query with all of the source events from all three campaigns, and all of the trigger events. This would make much better use of their budget, as well as enable “global attribution”, where we can avoid double counting.

To be clear, I understand this is a significant departure from how things work today. Today Facebook ads manager shows just an FB view of things. In an IPA world, it would not be possible to show them this. It wouldn’t be an efficient use of their privacy budget. It would be much more similar to the mobile app ecosystem where advertisers utilize 3rd party “mobile measurement partners” that give them a unified view across all their ad buying channels, preferring to view reporting there and eschewing platform-specific reporting channels.

**Comment by csharrison** [comment]
> In the event an advertiser wants to evaluate the relative performance of 3 campaigns (which they might have purchased from different ad-sellers) I assume that they would NOT issue three separate queries as you’ve shown. This would wind up hitting their differential privacy budget three times for the same set of trigger events. They’d be far better off running a single query with all of the source events from all three campaigns, and all of the trigger events. This would make much better use of their budget, as well as enable “global attribution”, where we can avoid double counting.

I think I might be missing something. Is this use-case possible to achieve with IPA:
> imagine there's a single advertiser selling a single product, and is running 3 different campaigns for it. They want to use IPA to measure the relative performance of their 3 campaigns.

i.e. I want a break-out that says:
Campaign1: led to 10 conversions
Campaign2: led to 15 conversions
Campaign3: led to 150 conversions

My thought from the doc was this is accomplished via carefully sending relevant source events, but let me know if / how you think this should be in done in IPA.

**Comment by eriktaubeneck** [comment]
> I think I might be missing something. Is this use-case possible to achieve with IPA:
> 
> > imagine there's a single advertiser selling a single product, and is running 3 different campaigns for it. They want to use IPA to measure the relative performance of their 3 campaigns.
> 
> i.e. I want a break-out that says: Campaign1: led to 10 conversions Campaign2: led to 15 conversions Campaign3: led to 150 conversions
> 
> My thought from the doc was this is accomplished via carefully sending relevant source events, but it seems like there is some other way this should be done. Here is the relevant piece from the doc:
> 
> > Note that source.example can use its own context and the context provided by trigger sites to group these queries into relevant sets. For example, if the source reports were a set of ad impressions, source.example could choose to run a query for a specific campaign, and only include trigger reports for items relevant to that campaign.
> 
> Now that is specific to a source query, but I assumed you'd do the same for trigger queries like the one I described.

Our wording in the doc may not have been super clear - there are two different cases to consider here.

The first case is the one you mention, you would want to issue a single query, with all the events. It would be something like the following SQL query:

```sql
select 
    source_event.campaign_id
  , count(trigger_event.event_id)
  , sum(trigger_event.value)
from
    source_events
    join trigger_events
    on <matchkeys and attribution logic>
group by
    source_event.campaign_id
```

The second case is where there are multiple distinct products involved, such as:

```
{
    (campaign_1a, campaign_1b, ...) : product_1,
    (campaign_2a, campaign_2b, ...): product_2,
    ...
}
```

In this case, since these queries can be constructed entirely independently, the advertiser running the query should be able to bifurcate them appropriately and run the same query as above, without having an effect on the results. In that case, having less data should be more efficient, and also not exhaust unnecessary privacy budget. It would also prevent the need for more complicated attribution logic in the MPC (since you'd only want attribution within that appropriate mapping.)

**Comment by csharrison** [comment]
Thanks @eriktaubeneck , I think I missed the piece where we can annotate source events by their relevant campaign ID. I wasn't sure if that was supported.

**Comment by csharrison** [comment]
I guess I will follow-up: how much extra information can we pack into the events? One of the benefits of creating queries as a  "bag of relevant events" is that we can use arbitrarily complex information to structure the queries. Once the splitting has to happen within the protocol though, it becomes harder, especially with MPC. Could you imagine us supporting many dimensions of features beyond campaign IDs in IPA?

**Comment by eriktaubeneck** [comment]
`campaign_id` is probably a bad name, TBH. `group_id` is better, and because the `source_event` is sent back in the context it was called, it shouldn't need to be bound to the event when called.  I've been thinking that one could assign the `group_id` after the fact (and even change it for two different queries.) The only restriction would be reaching the set grouping threshold (+DP noise).

I haven't thought about assigning multiple `group_id`s to a single event within a single query, and I have the same intuition around increasing complexity in the MPC. Are there scenarios where this would be important to have multiple in a single query (as opposed to sequential, independent queries?)

**Comment by csharrison** [comment]
If we are allowed to set `group_id` after the fact (i.e. at query time) it resolves my concern, i.e. that it should be possible to make complicated queries from arbitrary data at query time. This is one of the downsides of the Attribution Reporting API in that you need to decide at report-creation time what the query needs to be, and the architecture of IPA seems well suited to solve that problem.

I will need to think a bit (and understand more about IPA) if there's any benefit to multiple `group_ids` per record in a single query. I do believe there will be a large benefit to batching queries, so if that can be supported in an efficient way it might not be necessary. For example, it would be great to have an answer for how we would support the [Criteo experiment](https://medium.com/criteo-engineering/results-from-the-criteo-adkdd-2021-challenge-50abc9fa3a6) in IPA (even theoretically). In that setting, we have each attributed event contributing separately to 171 buckets.

**Comment by benjaminsavage** [comment]
I agree with @eriktaubeneck. Let me just elaborate a bit:

Since the source_event is generated in-context, you'll know all the relevant queries in which you might like to use it. Perhaps a country breakdown, an age-and-gender breakdown, a placement breakdown, etc. 

I'm assuming that the "group_id" can be added server-side at will, and the events can be utilized in multiple queries.

So an advertiser who wishes to get multiple breakdowns for their conversions would have to decide how much of their privacy budget to spend per-breakdown, then could issue multiple queries using the same events. 

As for could it contribute to 171 buckets? I think the answer depends entirely on the privacy budget.

**Comment by csharrison** [comment]
> I agree that this seems like a good goal worth shooting for. At the moment, I'd be happy to start with something simple (like last touch or even credit over N touches), but with the flexibility to get more complicated.

To be clear, this example is using last-touch attribution. It's just that we want to sum up not just campaign counts but also other features, so we can know things like "how many attributed events had feature X", "how many attributed events had feature Y", etc.

@benjaminsavage yes, privacy budget :) Actually this batching queries thing gives you a super power, budget wise, which is exactly what the Criteo takes advantage of in their competition. See also [this thread](https://github.com/WICG/conversion-measurement-api/issues/249)

**Comment by benjaminsavage** [comment]
A quote from that thread:

> We believe that there are great benefits in aggregating reports from multiple source_site (or attribution_destination, depending on the use case) in a single request, to lower the overall level of noise.

I agree. I would really like for IPA to support queries where the source_events span multiple source sites. I think this is a key use-case for ad-networks that show ads across the open web. We discuss this possible extension in our IPA proposal in the "[business privacy grain](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit#heading=h.3fuxnvaozwm4)" section. It's really hard though, and we haven't yet worked through all the issues with this. In particular, it requires careful design to ensure a malicious helper node cannot violate the "Vegas Rule".

Reading through that thread, the use-case is really about training ML, not reporting. Rather than trying to get hundreds of independent breakdowns out of the API, it would probably be more efficient (from a DP perspective) to just train an ML model in MPC, and emit a trained model (with DP noise added). We allude to this as a possible future extension: [link](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit#heading=h.82taoxx5dmqm). This would have the added benefit of being able to model the interaction effects between these features.

**Comment by csharrison** [comment]
> I agree. I would really like for IPA to support queries where the source_events span multiple source sites. I think this is a key use-case for ad-networks that show ads across the open web. We discuss this possible extension in our IPA proposal in the "[business privacy grain](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit#heading=h.3fuxnvaozwm4)" section. It's really hard though, and we haven't yet worked through all the issues with this. In particular, it requires careful design to ensure a malicious helper node cannot violate the "Vegas Rule".

I don't think business privacy grain is necessary here. The thread there is about combining reports across publishers e.g. for a given advertiser. My understanding is that IPA supports this by default (and uses, in that example, the advertiser privacy unit).

> Reading through that thread, the use-case is really about training ML, not reporting. Rather than trying to get hundreds of independent breakdowns out of the API, it would probably be more efficient (from a DP perspective) to just train an ML model in MPC, and emit a trained model (with DP noise added). We allude to this as a possible future extension: [link](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit#heading=h.82taoxx5dmqm). This would have the added benefit of being able to model the interaction effects between these features.

Yes, I used this mostly as an example, to understand the limitations of IPA. Obviously if we can train models directly in IPA it will probably be more efficient, but supporting the Criteo competition setting is a decent litmus test on how powerful the reporting use-case is. As far as I understand, supporting a setting like this could allow us to do logistic regression in a pretty privacy-efficient way.

**Comment by benjaminsavage** [comment]
Hi @juanli16 - and thanks for the question!

Our current thinking is that the API caller would provide some kind of "zero knowledge proof" along with each trigger event, proving that the trigger value lies within a given range.

The actual range would also need to be an API param, as the MPC would need to add noise proportional to that value. This param value would need to align with the zero knowledge proofs provided.

I would not imagine adding any local differential privacy.

**Comment by eriktaubeneck** [comment]
The solution @benjaminsavage references is the one [presented in PRIO](https://crypto.stanford.edu/prio/). It's important to clarify though, this requires a global bound on trigger values (for example we could pick the range [0,2^16) and use 16 bits.) Depending on the cryptographic details of the MPC, we may also be able to limit that just by limiting the secret shared values (and avoid the extra work of a zero-knowledge proof.)

To address the main point of your question, @juanli16, the range is in fact revealed to the aggregators (it is a global constant), but individual values are not revealed to the aggregators. And because the range is a known quantity, the distribution from which to draw the DP noise is also known.

**Comment by juanli16** [comment]
Thanks @benjaminsavage and @eriktaubeneck for the clarifications! I was starting to arrive to the same conclusion after inspecting the current state of this repo: [raw-ipa](https://github.com/martinthomson/raw-ipa). I do have 2 follow up questions if you would indulge me.

1. With the global limit on what a trigger values can be, how do we deal with trigger values that will be out of bound? Would it make sense for the browser that is generating the trigger event with an out of bound trigger value to split the event in multiple individual trigger events, each with a trigger value that satisfies the global bound, but still sums up to the original value (basically additive secret share)?

2. If the trigger values are very small that represent for example the number of clicks/app installs, etc. Wouldn't it be possible for the differential private noise computed based on the global bound (much larger than the trigger values) to skew the final aggregate value and lower the over accuracy of the result?

**Comment by benjaminsavage** [comment]
A few thoughts:

1. We might not have to select the "trigger value" in the browser. It might be possible to select it later, server-side. This would include some amount of work to split it up into secret shares and compute a ZKP server-side. 
2. I suspect the bound on trigger values will be scoped to the Query. So if you run 10 different queries, you could provide a different bound for each. This should help with your problem #2. You could run one query which is just counting app-installs, and each trigger event would just have a trigger value of 1. You could run a separate query which is totalling sales, and use a much higher bound.
3. I do not think we will have to resort to hacks like splitting an event up into multiple events. However, there *will* need to be a (per-query specified) total limit, of the total amount a single user can contribute to the value. This is a separate limit in addition to the per trigger-event limit on the trigger value. This is necessary to provide a differential privacy guarantee. So unfortunately, if one user makes a ton of huge purchases, their total contribution will be capped. Again - as this is a per query param, you can find your own optimal tradeoff between a high value (less capping, but more DP noise added), or a low value (more capping, but less DP noise added).

**Comment by alextcone** [comment]
Is this still where we should open issues on IPA?

**Comment by benjaminsavage** [comment]
Hi @alextcone - IPA related issues can be filed here: https://github.com/patcg-individual-drafts/ipa/issues
